fennel/internal_lib/to_proto/to_proto.py

```diff
@@ -1104,6 +1113,78 @@ def _renames_to_proto(renames: Optional[Dict[str, str]]) -> Mapping[str, str]:
     return renames
 
 
+def get_sampling_rate(sample: Optional[Union[float, Sample]]):
+    if isinstance(sample, float):
+        if sample < 0 or sample > 1:
+            raise ValueError("Sample rate should be between 0 and 1")
+        return sample
+    if isinstance(sample, Sample):
+        if sample.rate < 0 or sample.rate > 1:
+            raise ValueError("Sample rate should be between 0 and 1")
+        return sample.rate
+    raise ValueError(f"Invalid sample type: {type(sample)}")
+
+
+def get_sampling_columns(
+    sample: Optional[Union[float, Sample]],
+    dataset: Dataset,
+    preproc: Optional[Dict[str, PreProcValue]],
+):
+    disallowed_columns = [dataset.timestamp_field]
+    if preproc is not None:
+        disallowed_columns.extend(list(preproc.keys()))
+
+    if isinstance(sample, Sample) and len(sample.using) != 0:
+        input_columns = sample.using
+        all_columns = [str(column) for column in dataset.fields]
+        for column in input_columns:
+            if column == dataset.timestamp_field:
+                raise ValueError(
+                    f"Timestamp column: {column} cannot be part of sampling columns"
+                )
+            if column in disallowed_columns:
+                raise ValueError(
+                    f"Column {column} is part of preproc so cannot be used for sampling"
+                )
+            if column not in all_columns:
+                raise ValueError(
+                    f"Column {column} is not part of dataset columns"
+                )
+        return sorted(input_columns)
+
+    if len(dataset.key_fields) != 0:
+        key_columns = dataset.key_fields
+        return sorted(
+            [
+                column
+                for column in key_columns
+                if column not in disallowed_columns
+            ]
+        )
+    else:
+        all_columns = [str(column) for column in dataset.fields]
+        return sorted(
+            [
+                column
+                for column in all_columns
+                if column not in disallowed_columns
```

Comment:
We should raise an error here given we are skipping some key fields to ask customer to add sampling columns

Reply:
throwing exception if all columns except timestamp are part of preproc.